File: benchmark/languages/go/work/day07/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func readGrid(filename string) []string {
	f, err := os.Open(filename)
	if err != nil {
		panic(err)
	}
	defer f.Close()
	var grid []string
	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		line := scanner.Text()
		if line != "" {
			grid = append(grid, line)
		}
	}
	return grid
}

// For Part 1: BFS/simulation of beams traveling downward.
// Each beam is identified by (col, row) where it starts traveling downward.
// When it hits a ^, it splits: two new beams start at (col-1, splitterRow) and (col+1, splitterRow)
// traveling downward.
// Count total number of unique splitter hits (activations).

func solve(grid []string) (int, int) {
	rows := len(grid)
	cols := len(grid[0])

	// Find S position (starting column, row 0)
	startCol := -1
	for c := 0; c < cols; c++ {
		if grid[0][c] == 'S' {
			startCol = c
			break
		}
	}

	// We simulate beams. A beam travels downward from a given column starting at a given row.
	// State: (col, startRow) - beam traveling down column col starting from row startRow.
	// We use BFS and track which splitters have been activated.

	// For Part 1: count unique splitter activations.
	// For Part 2: count number of timelines = number of leaf paths.

	// Think of it as a DAG:
	// - Root: beam at (startCol, row 1) going down (after S at row 0)
	//   Actually beam starts at row 0 from S, going down.
	// 
	// For each beam (col, startRow going down):
	//   Scan rows from startRow downward in column col.
	//   If we hit a '^' at (col, r):
	//     - This splitter is activated (count it for part 1 if not yet counted)
	//     - Spawn two beams: (col-1, r) and (col+1, r) going down
	//     - (if col-1 or col+1 out of bounds, that beam just exits)
	//   If we exit the grid without hitting a splitter: this is a leaf (for part 2, +1 timeline)

	// For Part 2: number of timelines = number of paths that reach a leaf (exit without splitting further).
	// We compute this recursively with memoization.
	// timelines(col, startRow) = 
	//   if next splitter is at row r:
	//     timelines(col-1, r) + timelines(col+1, r)   [but handle out-of-bounds: OOB = 1 timeline]
	//   else (no splitter found):
	//     1

	// But wait: if two beams arrive at the same splitter from the same direction,
	// in Part 1 the splitter counts once (beams merge).
	// In Part 2, beams are independent timelines so they don't merge.

	// Actually re-reading Part 2: "A tachyon particle takes both the left and right path"
	// and we count distinct timelines. If two paths converge to the same splitter,
	// they still count as separate timelines past that point.
	// So for Part 2, timelines(col, startRow) with memoization where the result
	// is the count of leaf exits from that beam state.

	// For Part 1: splitters activated = unique (col, row) splitter positions that are hit.

	// Let's find all splitter positions: grid[r][c] == '^'
	// For a beam at column col starting at row startRow going down:
	//   find smallest r >= startRow such that grid[r][col] == '^'
	//   (Also S doesn't block beams, it's just the starting marker)

	// Memoization for Part 2:
	// memo[col][startRow] = number of leaf timelines from beam (col, startRow going down)
	memo := make([][]int, cols)
	computed := make([][]bool, cols)
	for c := 0; c < cols; c++ {
		memo[c] = make([]int, rows)
		computed[c] = make([]bool, rows)
	}

	// Set of activated splitters for Part 1
	activated := make(map[[2]int]bool)

	// Find next splitter in column col at or below row startRow
	// Returns row of splitter, or -1 if none
	nextSplitter := func(col, startRow int) int {
		for r := startRow; r < rows; r++ {
			if grid[r][col] == '^' {
				return r
			}
		}
		return -1
	}

	// We need to do two things:
	// 1. Simulate all reachable beam states (BFS) to find activated splitters (Part 1)
	// 2. Count timelines (Part 2) via memoized recursion

	// Actually for Part 1, we also need to know which splitters are reachable.
	// Let's use BFS for Part 1, then memoized recursion for Part 2 restricted to reachable splitters.

	// Wait - for Part 2, the timelines function doesn't care about "reachable" from a global perspective;
	// it's just the tree structure. Every beam that reaches a splitter spawns 2 children.
	// If col-1 < 0 or col+1 >= cols, that child just exits (1 timeline).

	// BFS for reachability (Part 1):
	type BeamState struct{ col, row int }
	visited := make(map[BeamState]bool)
	queue := []BeamState{{startCol, 0}}
	visited[BeamState{startCol, 0}] = true

	for len(queue) > 0 {
		b := queue[0]
		queue = queue[1:]

		r := nextSplitter(b.col, b.row)
		if r == -1 {
			// beam exits, no split
			continue
		}
		// Splitter at (b.col, r) is activated
		activated[[2]int{b.col, r}] = true

		// Spawn children
		children := []BeamState{}
		if b.col-1 >= 0 {
			children = append(children, BeamState{b.col - 1, r})
		}
		if b.col+1 < cols {
			children = append(children, BeamState{b.col + 1, r})
		}
		for _, child := range children {
			if !visited[child] {
				visited[child] = true
				queue = append(queue, child)
			}
		}
	}

	part1 := len(activated)

	// Part 2: count timelines via memoized recursion
	// timelines(col, startRow) = number of leaf exits
	var countTimelines func(col, startRow int) int
	countTimelines = func(col, startRow int) int {
		if col < 0 || col >= cols {
			return 1 // exited left/right
		}
		if computed[col][startRow] {
			return memo[col][startRow]
		}
		computed[col][startRow] = true

		r := nextSplitter(col, startRow)
		if r == -1 {
			// no splitter found, beam exits bottom
			memo[col][startRow] = 1
			return 1
		}
		// split at row r: children are (col-1, r) and (col+1, r)
		left := 1
		right := 1
		if col-1 >= 0 {
			left = countTimelines(col-1, r)
		}
		if col+1 < cols {
			right = countTimelines(col+1, r)
		}
		result := left + right
		memo[col][startRow] = result
		return result
	}

	part2 := countTimelines(startCol, 0)

	return part1, part2
}

func main() {
	grid := readGrid("input.txt")

	part := ""
	if len(os.Args) > 1 {
		part = os.Args[1]
	}

	p1, p2 := solve(grid)

	switch part {
	case "1":
		fmt.Printf("Part 1: %d\n", p1)
	case "2":
		fmt.Printf("Part 2: %d\n", p2)
	default:
		fmt.Printf("Part 1: %d\n", p1)
		fmt.Printf("Part 2: %d\n", p2)
	}
}

File: benchmark/languages/c/work/day07/main.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAXROWS 300
#define MAXCOLS 300

char grid[MAXROWS][MAXCOLS+1];
int rows, cols;

// splitter_hit[r][c] = 1 if the splitter at (r,c) has been activated
int splitter_hit[MAXROWS][MAXCOLS];

typedef struct {
    int row, col;
} Beam;

// Queue for BFS of beam starting positions
Beam queue[MAXROWS * MAXCOLS * 4];
int qhead, qtail;

// visited[r][c] = 1 if a beam starting at (r,c) has been processed
int visited[MAXROWS][MAXCOLS];

// next_split[r][c] = row of next ^ at column c at or below row r, -1 if none
int next_split[MAXROWS+1][MAXCOLS];

void precompute_next_split() {
    for (int c = 0; c < cols; c++) {
        next_split[rows][c] = -1;
        for (int r = rows-1; r >= 0; r--) {
            if (grid[r][c] == '^') {
                next_split[r][c] = r;
            } else {
                next_split[r][c] = next_split[r+1][c];
            }
        }
    }
}

int find_start(int *s_row, int *s_col) {
    for (int r = 0; r < rows; r++) {
        for (int c = 0; c < cols; c++) {
            if (grid[r][c] == 'S') {
                *s_row = r; *s_col = c;
                return 1;
            }
        }
    }
    return 0;
}

// Part 1: Simulate beams, count unique splitters activated
// Each beam starts at (row, col) moving downward.
// When it hits a splitter, that splitter is activated (count if first time).
// Spawn beams at (hit_row, c-1) and (hit_row, c+1) if not yet visited.
long long simulate_part1(int s_row, int s_col) {
    memset(splitter_hit, 0, sizeof(splitter_hit));
    memset(visited, 0, sizeof(visited));
    qhead = qtail = 0;
    
    queue[qtail++] = (Beam){s_row, s_col};
    visited[s_row][s_col] = 1;
    
    long long splits = 0;
    
    while (qhead < qtail) {
        Beam b = queue[qhead++];
        int r = b.row, c = b.col;
        
        int hit = next_split[r][c];
        
        if (hit < 0) {
            // Beam exits, no split
            continue;
        }
        
        // Beam hits splitter at (hit, c)
        if (!splitter_hit[hit][c]) {
            splitter_hit[hit][c] = 1;
            splits++;
        }
        
        // Spawn left beam at (hit, c-1)
        int lc = c - 1;
        if (lc >= 0 && !visited[hit][lc]) {
            visited[hit][lc] = 1;
            queue[qtail++] = (Beam){hit, lc};
        }
        
        // Spawn right beam at (hit, c+1)
        int rc2 = c + 1;
        if (rc2 < cols && !visited[hit][rc2]) {
            visited[hit][rc2] = 1;
            queue[qtail++] = (Beam){hit, rc2};
        }
    }
    
    return splits;
}

// Part 2: Count timelines using memoization.
// memo[r][c] = number of timelines for a single particle starting at (r,c) going down.
// timelines(r,c) = 
//   if no splitter below from (r,c): 1 (exits, 1 timeline)
//   if splitter at (hit,c):
//     left contribution: (lc >= 0) ? memo[hit][lc] : 1  (exits immediately = 1)
//     right contribution: (rc2 < cols) ? memo[hit][rc2] : 1
//     total: left + right
//
// We need memoization because the same starting position can be reached from multiple paths.
// Use iterative DFS with explicit stack to avoid C stack overflow on large inputs.

long long memo[MAXROWS][MAXCOLS];
char computed[MAXROWS][MAXCOLS];

typedef struct { int r, c, phase; } Frame;
Frame dfs_stack[MAXROWS * MAXCOLS * 2];

long long compute_timelines(int s_row, int s_col) {
    memset(computed, 0, sizeof(computed));
    for (int r = 0; r < rows; r++)
        for (int c = 0; c < cols; c++)
            memo[r][c] = -1;
    
    int top = 0;
    dfs_stack[top++] = (Frame){s_row, s_col, 0};
    
    while (top > 0) {
        Frame *fr = &dfs_stack[top-1];
        int r = fr->r, c = fr->c;
        
        if (computed[r][c]) {
            top--;
            continue;
        }
        
        int hit = next_split[r][c];
        
        if (hit < 0) {
            // No splitter, beam exits = 1 timeline
            memo[r][c] = 1;
            computed[r][c] = 1;
            top--;
            continue;
        }
        
        int lc = c - 1, rc2 = c + 1;
        
        if (fr->phase == 0) {
            fr->phase = 1;
            // Push dependencies if not computed
            int pushed = 0;
            if (rc2 < cols && !computed[hit][rc2]) {
                dfs_stack[top++] = (Frame){hit, rc2, 0};
                pushed = 1;
            }
            if (lc >= 0 && !computed[hit][lc]) {
                dfs_stack[top++] = (Frame){hit, lc, 0};
                pushed = 1;
            }
            if (pushed) continue;
        }
        
        // Dependencies are ready
        long long left_t = (lc >= 0) ? memo[hit][lc] : 1;
        long long right_t = (rc2 < cols) ? memo[hit][rc2] : 1;
        
        memo[r][c] = left_t + right_t;
        computed[r][c] = 1;
        top--;
    }
    
    return memo[s_row][s_col];
}

int main(int argc, char *argv[]) {
    int part = 0;
    if (argc > 1) part = atoi(argv[1]);
    
    FILE *f = fopen("input.txt", "r");
    if (!f) { fprintf(stderr, "Cannot open input.txt\n"); return 1; }
    
    rows = 0;
    while (fgets(grid[rows], MAXCOLS+1, f)) {
        int len = strlen(grid[rows]);
        while (len > 0 && (grid[rows][len-1] == '\n' || grid[rows][len-1] == '\r'))
            grid[rows][--len] = '\0';
        if (len > 0) {
            cols = len;
            rows++;
        }
    }
    fclose(f);
    
    precompute_next_split();
    
    int s_row = 0, s_col = 0;
    find_start(&s_row, &s_col);
    
    long long ans1 = 0, ans2 = 0;
    
    if (part == 0 || part == 1) {
        ans1 = simulate_part1(s_row, s_col);
        printf("Part 1: %lld\n", ans1);
    }
    
    if (part == 0 || part == 2) {
        ans2 = compute_timelines(s_row, s_col);
        printf("Part 2: %lld\n", ans2);
    }
    
    return 0;
}
